feat(Codecov): add storage solution

adrianviquez opened the pull request:
This PR adds a context to the Codecov product to hold state and functionality to save/get data to/from the url and local storage. This context is used in the CodecovProvider that will be the main wrapper for all things Codecov. This is the preferred approach over [this pr](https://github.com/getsentry/sentry/pull/89818).

Notes
-
- When I name things "Codecov" here, I'm referring it as a product and not as the company. So thinking of it as a "feature" or "product" such as "replay" or "issues". It might be unnecessary for some cases (like the persistence/url), but I think it makes sense for the provider/context pieces. Open to thoughts here!
- `CodecovProvider`: this is the main code that will act as a wrapper for everything in the Codecov product. It instantiates and initializes the CodecovContext component, as well as sets functions to manipulate the context.
- `CodecovContext`: the context itself.

Thanks!

adrianviquez committed:
Add Codecov context and functions to store in URL + local storage

adrianviquez committed:
Merge remote-tracking branch 'origin/master' into adrian/add-storage-solution

adrianviquez commented on static/app/components/codecov/container/persistence.tsx:
```diff
@@ -0,0 +1,83 @@
+ * 2) Attempts to parse the data.
+ * 3) Validates and returns the data.
+ */
+export function getCodecovDataFromLocalStorage(
```
You'll see both methods here are get/setCodecovData... . As stated in the description, I'm thinking of Codecov here more as a product, so the idea with these helper methods is to help choose the correct key and decode the data that's needed according to what this method expects. 

Thinking it through though, I think we could make it more generic and be something like this

``` ts
export function getDataFromLocalStorage<T>(
  localStorageKey: string,
): T | null {
  const value = localStorage.getItem(localStorageKey);

  if (!value) {
    return null;
  }

  let decoded: T;

  try {
    decoded = JSON.parse(value);
  } catch (err) {
    captureException(err);
    return null;
  }

  return {...decoded}
}
```

And elevate the key to the caller. And do something similar for the setter. Wdyt?

adrianviquez replied:
I actually went back and did this, so new fn follows this. Also did that for setting data to local storage and query data from the url

adrianviquez commented on static/app/components/codecov/container/persistence.tsx:
```diff
@@ -0,0 +1,83 @@
+const CODECOV_PRODUCT = 'codecov';
+
+// Types
+type CodecovObjectInLocalStorage = {
```
Currently redundant; could extend from the type in the Context component if this eventually grows or simply reuse it. Leaning towards reusing the original instead of extending.

adrianviquez commented on static/app/components/codecov/container/persistence.tsx:
```diff
@@ -0,0 +1,83 @@
+export function getCodecovDataFromLocalStorage(
+  orgSlug: string
+): CodecovObjectInLocalStorage | null {
+  // 1)
```
Trying this approach to comment functionality, not sure how I feel but wanted opinions

adrianviquez committed:
Refactor url + local storage methods to be more generic

adrianviquez committed:
simplify functionality with query params and local storage

adrianviquez committed:
Merge remote-tracking branch 'origin/master' into adrian/add-storage-solution

adrianviquez committed:
Update param type definition

JonasBa commented on static/app/components/codecov/container/container.tsx:
```diff
@@ -0,0 +1,104 @@
+import {useNavigate} from 'sentry/utils/useNavigate';
+import useOrganization from 'sentry/utils/useOrganization';
+
+// Types
```
![image](https://github.com/user-attachments/assets/1dbf40c7-28fb-4451-87e5-2e0a99c203ad)

Types should explain why, not what :)

adrianviquez replied:
Will adjust all comments

JonasBa commented on static/app/components/codecov/container/container.spec.tsx:
This shouldn't have been checked in

JonasBa commented on static/app/components/codecov/container/container.tsx:
```diff
@@ -0,0 +1,104 @@
+  integratedOrg: string | null;
+  repository: string | null;
+};
+export type CodecovContextSetterTypes = {
+  setContextState: React.Dispatch<React.SetStateAction<CodecovContextTypes>>;
+  updateSelectorData: (value: Partial<CodecovContextTypes>) => void;
+};
```
```suggestion
export type CodecovContextSetterTypes = {
  setContextState: React.Dispatch<React.SetStateAction<CodecovContextTypes>>;
  updateSelectorData: (value: Partial<CodecovContextTypes>) => void;
};
```

Why do you need both here? If someone calls setContextState, it means it wont end up updating query params, potentially detaching state.

adrianviquez replied:
Yeah true, I don't need to expose the setContextState fn, that should be private to the context itself. Left it there from the original implementation but no need to have it

JonasBa commented on static/app/components/codecov/container/container.tsx:
```diff
@@ -0,0 +1,104 @@
+import useOrganization from 'sentry/utils/useOrganization';
+
+// Types
+export type CodecovContextTypes = {
```
```suggestion
export type CodecovContext = {
```
There is no need to annotate this any further

JonasBa commented on static/app/components/codecov/container/container.tsx:
```diff
@@ -0,0 +1,104 @@
+  integratedOrg: string | null;
+  repository: string | null;
+};
+export type CodecovContextSetterTypes = {
```
Why do you export these separately?

JonasBa commented on static/app/components/codecov/container/container.tsx:
```diff
@@ -0,0 +1,104 @@
+};
+type CodecovUrlParams = 'repository' | 'integratedOrg' | 'branch' | 'codecovPeriod';
+
+type CodecovProviderProps = {
+  children?: React.ReactNode;
+};
```
```suggestion
type CodecovProviderProps = {
  children: NonNullable<React.ReactNode>;
};
```

What would be the use case for using this provider without children?

JonasBa commented on static/app/components/codecov/container/container.tsx:
```diff
@@ -0,0 +1,104 @@
+  children?: React.ReactNode;
+};
+
+// Constants
```
```suggestion
```

JonasBa commented on static/app/components/codecov/container/container.tsx:
```diff
@@ -0,0 +1,104 @@
+};
+
+// Constants
+const DEFAULT_CODECOV_CONTEXT_VALUES: CodecovContextTypes = {
+  repository: null,
+  integratedOrg: null,
+  branch: null,
+  codecovPeriod: null,
+};
```
If they are all null, then why even have a default value?

JonasBa commented:
@adrianviquez I would suggest to simplify this to not use React state and be resilient to query param updates outside of the provider.

```tsx
function CodecovQueryParamsProvider(props){
  const location = useLocation();
  const localStorageState = useLocalStorage();

  const params = {
    key: location.query.key || localStorageState.key || default value
    ...
  }

  return <Provider value={params}>{props.children}</Provider>
}
```

The only part left then is to decide how to sync query to localStorage, which you could probably just do from inside an effect inside this provider

adrianviquez committed:
Merge remote-tracking branch 'origin/master' into adrian/add-storage-solution

adrianviquez commented on static/app/components/codecov/context/context.tsx:
```diff
@@ -0,0 +1,16 @@
+import {createContext, useContext} from 'react';
+
+import type {
+  CodecovContextSetterTypes,
+  CodecovContextTypes,
+} from 'sentry/components/codecov/container/container';
```
I renamed these to codecovParamsProvider.tsx and codecovContext.tsx. I normally let the folder help me with understanding it, so like codecov/components/context helps me understand it's a codecovContext, but it is true it can be clearer to just name it codecov<X>

adrianviquez committed:
Refactor logic to promote url as state

adrianviquez commented:
Ty for the feedback @JonasBa. I refactored the component to encourage using url as the state. I believe I addressed all of your suggestions, apologies if I missed one.

I created a `updateSelectorData` function that standardizes how to update the url. I believed your original suggestion was for each component to be responsible for this logic, but I thought standardizing how to do it in this class made sense. I'm flexible about this though if you think the reusability value isn't too high - I also noticed after the fact you said exposing data and setters tends to be a react anti-pattern, so open to adjusting this.

Please let me know what you think about these changes, thanks for all the care and feedback 🙇

adrianviquez committed:
Adjust updateSelectorData input type

JonasBa commented on static/app/components/codecov/container/codecovParamsProvider.tsx:
```diff
@@ -0,0 +1,72 @@
+
+  const params: CodecovContextData = {
+    repository:
+      decodeURIComponent(location.query.repository as string).trim() ||
```
Why is .trim() required?

adrianviquez replied:
In case the values I get have whitespace in a future. Technically I control the input so I know it isn't happening today, so this is more of a future proof case. 

Technically adding complexity now for robustness later, so I'm okay getting rid of it today and adding it later if we anticipate strings with whitespaces

JonasBa replied:
Why would it get a whitespace, and why would the whitespace be invalid?

adrianviquez replied:
I suppose there wouldn't be for now since the value is controlled by data we manipulate, not user data. And even if so, it should be okay.

JonasBa commented on static/app/components/codecov/container/codecovParamsProvider.tsx:
```diff
@@ -0,0 +1,72 @@
+  );
+
+  useEffect(() => {
+    const repository =
+      typeof location.query.repository === 'string' ? location.query.repository : null;
+    const integratedOrg =
+      typeof location.query.integratedOrg === 'string'
+        ? location.query.integratedOrg
+        : null;
```
Do you need to check for empty string?

JonasBa commented on static/app/components/codecov/container/codecovParamsProvider.tsx:
```diff
@@ -0,0 +1,72 @@
+    repository:
+      decodeURIComponent(location.query.repository as string).trim() ||
+      localStorageState.repository ||
+      null,
```
Should there be any defaults for this? The way this reads right now is as if null is a valid value

adrianviquez replied:
I was thinking of the nulls as a legitimate placeholder value until a person chooses a value for the selector. So if the value is null the selector would say like "choose repo" or something. Would you recommend against this and instead have the value be undefined or not sent via the context altogether?

JonasBa replied:
That makes sense, but I would make sure to document it. You are probably going to run into issues down the line where the fetch calls are being made with null values (unless you explicitly check obviously).

adrianviquez replied:
Sounds good, left a comment explaining this 👌

adrianviquez committed:
fix: adjust logic to populate selector from local storage when url data is missing

adrianviquez committed:
Merge remote-tracking branch 'origin/master' into adrian/add-storage-solution

adrianviquez committed:
fix: get rid of unecessary trim()

adrianviquez committed:
ignore commited file

adrianviquez committed:
Add comment explaining null placeholder values to context data

getsantry[bot] committed:
:hammer_and_wrench: apply pre-commit fixes

JonasBa commented on static/app/components/codecov/container/codecovParamsProvider.tsx:
```diff
@@ -0,0 +1,75 @@
+
+  const location = useLocation();
+  const [localStorageState, setLocalStorageState] = useLocalStorageState(
+    `codecov-selection:${orgSlug}`,
```
```suggestion
    `codecov-selection:${organization.slug}`,
```

JonasBa commented on static/app/components/codecov/container/codecovParamsProvider.tsx:
```diff
@@ -0,0 +1,75 @@
+
+export default function CodecovQueryParamsProvider({children}: CodecovProviderProps) {
+  const organization = useOrganization();
+  const orgSlug = organization.slug;
```
```suggestion
```

JonasBa reviewed (approved):
Minor things left, but overall looks good, thanks for the patience

adrianviquez committed:
"adjust pr from feedback for better type safety"

adrianviquez committed:
Merge branch 'master' into adrian/add-storage-solution
